### PR TITLE
replace exporter timeout with receiver timeout

### PR DESCRIPTION
**Pull request by moh-osman3**
This PR is a followup to https://github.com/f5/otel-arrow-adapter/pull/207

Instead of the timeout happening in the client using `max_stream_lifetime` the timeout is moved to the server.  

1.) The timeout in the server is `max_connection_age - 2s`, because sufficiently small grace period could cause gRPC to interrupt the connection before the server is able to send a signal to the client to close the stream. So I am subtracting a couple seconds to give the server extra time to send the shutdown signal to the client, otherwise shutdown may be ungraceful.

2.) The client could potentially get two signals from the server, an `error` and `end_of_lifetime` signal. This might take a couple of iterations for `client.stream.write` to process both signals, but both should eventually be processed. This can be seen in the added unit tests where an extra call to `SendAndWait` is needed to see the error for stream restart.

### CODE REVIEW

File: collector/gen/exporter/otlpexporter/internal/arrow/stream.go [hunk 1]
```diff
@@ -350,20 +346,20 @@ func (s *Stream) read(_ context.Context) error {
...
 
+		if resp.EndOfLifetime {
+			close(s.closed)
+		}
+
 		if err = s.processBatchStatus(resp); err != nil {
 			return fmt.Errorf("process: %w", err)
+		} else if resp.EndOfLifetime {
+			// no errors but received signal from server to shutdown
+			return nil
```
Comment (jmacd) [inline collector/gen/exporter/otlpexporter/internal/arrow/stream.go:361]: I think we should keep reading until all the requests that have been sent have responses, so the condition maybe something like

```
if resp.EndOfLifetime && s.numWaiters() == 0 {
   return nil
}
```

with 

```
func (s *Stream) numWaiters() int {
	s.lock.Lock()
	defer s.lock.Unlock()

	return len(s.waiters)
}
```

File: collector/gen/exporter/otlpexporter/internal/arrow/stream.go [hunk 2]
```diff
@@ -64,6 +57,8 @@ type Stream struct {
 
 	// waiters is the response channel for each active batch.
 	waiters map[int64]chan error
+
+	closed chan int
```
[the remark this answers is not in the document]
Comment (moh-osman3) [reply collector/gen/exporter/otlpexporter/internal/arrow/stream.go]: Changed!

File: collector/gen/exporter/otlpexporter/internal/arrow/stream.go [hunk 3]
```diff
@@ -350,20 +346,20 @@ func (s *Stream) read(_ context.Context) error {
...
-			status, ok := status.FromError(err)
-			if ok && status.Message() == "EOF" && status.Code() == codes.Unknown {
-				return nil
-			}
 			// Note: do not wrap, contains a Status.
 			return err
 		}
 
+		if resp.EndOfLifetime {
+			close(s.closed)
```
[the remark this answers is not in the document]
Comment (moh-osman3) [reply collector/gen/exporter/otlpexporter/internal/arrow/stream.go]: Ahh good point, thanks for pointing this out! Should be fixed now

File: collector/gen/exporter/otlpexporter/internal/arrow/stream.go (hunk 1 above)
Comment (moh-osman3) [reply collector/gen/exporter/otlpexporter/internal/arrow/stream.go:365]: Makes sense to me, fixed!

File: collector/examples/synthesize/record.yaml [hunk 4]
```diff
@@ -1,8 +1,33 @@
...
+        endpoint: 127.0.0.1:4000
+        keepalive:
+          server_parameters:
+            max_connection_age: 1s
+            max_connection_age_grace: 1s
+      arrow:
+        disabled: false
+  otlp:
+    protocols:
+      grpc:
```
[the remark this answers is not in the document]
Comment (moh-osman3) [reply collector/examples/synthesize/record.yaml]: Thanks for the review! I will move settings to replay.yaml. I'm actually noticing a bug in this code - When the client restarts it doesn't seem like it gets brand new keepalive settings. 

e.g. if `MaxConnectionAge=20s` and the server timeout is `18s`. Then after 18s, the client receives shutdown and restarts, but then 2s later the server returns NO_ERROR. I would have expected a new `MaxConnectionAge` but it seems that the grpc disconnects ungracefully after the remainder of the last maxconnectionage? I think this means the connection is not being reset for some reason

**Pull request closed without merging**